stdlib/public/SwiftShims/RefCount.h

```diff
@@ -20,24 +20,503 @@
+
+  DEINITING with side table
+  Weak variable load returns nil. 
+  Weak variable store stores nil.
+  canBeFreedNow() is always false, so it never transitions directly to DEAD.
+  Everything else is the same as DEINITING.
+
+  DEINITED without side table
+  deinit() has completed but there are unowned references outstanding.
+  Strong variable operations can't happen.
+  Unowned variable store can't happen.
+  Unowned variable load halts in swift_abortRetainUnowned().
+  Weak variable operations can't happen.
+  When the unowned RC reaches zero, the object is freed and it becomes DEAD.
+
+  DEINITED with side table
+  Weak variable load returns nil.
+  Weak variable store can't happen.
+  When the unowned RC reaches zero, the object is freed, the weak RC is 
+    decremented, and the object becomes FREED.
+  Everything else is the same as DEINITED.
+
+  FREED without side table
+  This state never happens.
+
+  FREED with side table
+  The object is freed but there are weak refs to the side table outstanding.
+  Strong variable operations can't happen.
+  Unowned variable operations can't happen.
+  Weak variable load returns nil.
+  Weak variable store can't happen.
+  When the weak RC reaches zero, the side table entry is freed and 
+    the object becomes DEAD.
+
+  DEAD
+  The object and its side table are gone.
+*/
+
+namespace swift {
+  struct HeapObject;
+  class HeapObjectSideTableEntry;
+}
+
+// FIXME: HACK: copied from HeapObject.cpp
+extern "C" LLVM_LIBRARY_VISIBILITY void
+_swift_release_dealloc(swift::HeapObject *object)
+  SWIFT_CC(RegisterPreservingCC_IMPL)
+    __attribute__((__noinline__, __used__));
+
+namespace swift {
+
+// FIXME: many `relaxed` in this file should be `consume`,
+// but (1) the compiler doesn't support `consume` directly,
+// and (2) the compiler promotes `consume` to `acquire` instead which
+// is overkill on our CPUs. But this might leave us vulnerable to
+// compiler optimizations that `relaxed` allows but `consume` ought not.
+#define relaxed std::memory_order_relaxed
+#define acquire std::memory_order_acquire
+#define release std::memory_order_release
+#define consume std::memory_order_consume
+
+
+// RefCountIsInline: refcount stored in an object
+// RefCountNotInline: refcount stored in an object's side table entry
+enum RefCountInlinedness { RefCountNotInline = false, RefCountIsInline = true };
+
+enum ClearPinnedFlag { DontClearPinnedFlag = false, DoClearPinnedFlag = true };
+
+enum PerformDeinit { DontPerformDeinit = false, DoPerformDeinit = true };
+
+
+// Basic encoding of refcount and flag data into the object's header.
+// FIXME: Specialize this for a 32-bit field on 32-bit architectures.
+template <RefCountInlinedness refcountIsInline>
+class RefCountBitsT {
+
+  friend class RefCountBitsT<RefCountIsInline>;
+  friend class RefCountBitsT<RefCountNotInline>;
+  
+  static const RefCountInlinedness Inlinedness = refcountIsInline;
+  
+  uint64_t bits;
```

Comment:
This is potentially 8-byte-aligned even on 32-bit architectures, which is an unfortunate layout and not what the compiler assumes.  (It isn't on ARM32, but IIRC it is on i386.)

You could make it packed and aligned(4) on 32-bit if you want.  You'll have to figure out how to do that in MSVC, though.